File: src/ObjectManager/ObjectManager/MaterialManager.cs
using OA.Configuration;
using OA.Materials;
using UnityEngine;
using UnityEngine.Rendering;

namespace OA
{
    public enum MatTestMode { Always, Less, LEqual, Equal, GEqual, Greater, NotEqual, Never }

    public enum MaterialType
    {
        None, Default, Standard, BumpedDiffuse, Unlit
    }

    public struct MaterialTextures
    {
        public string MainFilePath;
        public string DarkFilePath;
        public string DetailFilePath;
        public string GlossFilePath;
        public string GlowFilePath;
        public string BumpFilePath;
    }

    public struct MaterialProps
    {
        public MaterialTextures Textures;
        public bool AlphaBlended;
        public BlendMode SrcBlendMode;
        public BlendMode DstBlendMode;
        public bool AlphaTest;
        public float AlphaCutoff;
        public bool ZWrite;
    }

    /// <summary>
    /// Manages loading and instantiation of materials.
    /// </summary>
    public class MaterialManager
    {
        BaseMaterial _material;

        public TextureManager TextureManager { get; }

        public MaterialManager(TextureManager textureManager)
        {
            TextureManager = textureManager;
            var game = BaseSettings.Game;
            switch (game.MaterialType)
            {
                case MaterialType.None: _material = null; break;
                case MaterialType.Default: _material = new DefaultMaterial(textureManager); break;
                case MaterialType.Standard: _material = new StandardMaterial(textureManager); break;
                case MaterialType.Unlit: _material = new UnliteMaterial(textureManager); break;
                default: _material = new BumpedDiffuseMaterial(textureManager); break;
            }
        }

        public Material BuildMaterialFromProperties(MaterialProps mp)
        {
            return _material.BuildMaterialFromProperties(mp);
        }

        private Material BuildMaterial()
        {
            return _material.BuildMaterial();
        }

        private Material BuildMaterialBlended(BlendMode sourceBlendMode, BlendMode destinationBlendMode)
        {
            return _material.BuildMaterialBlended(sourceBlendMode, destinationBlendMode);
        }

        private Material BuildMaterialTested(float cutoff = 0.5f)
        {
            return _material.BuildMaterialTested(cutoff);
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/ObjectManager/MaterialManager.swift
//
//  MaterialManager.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

//public enum MatTestMode { Always, Less, LEqual, Equal, GEqual, Greater, NotEqual, Never }

import SceneKit

public enum MaterialType {
    case standard
}

public enum BlendMode: UInt8 {
    case one
}

public enum MatTestMode: UInt8 {
    case one
}

public struct MaterialTextures {
    public var mainFilePath: String?
    public var darkFilePath: String?
    public var detailFilePath: String?
    public var glossFilePath: String?
    public var glowFilePath: String?
    public var bumpFilePath: String?
}

public struct MaterialProps: Hashable {
    public static func ==(lhs: MaterialProps, rhs: MaterialProps) -> Bool {
        return lhs.textures.mainFilePath == rhs.textures.mainFilePath
    }
    
    public var hashValue: Int { return 0 }
    public var textures: MaterialTextures!
    public var alphaBlended: Bool!
    public var srcBlendMode: BlendMode!
    public var dstBlendMode: BlendMode!
    public var alphaTest: Bool!
    public var alphaCutoff: Float!
    public var zwrite: Bool!
}

public class MaterialManager {
    var _material: BaseMaterial

    private(set) public var textureManager: TextureManager

    init(textureManager: TextureManager) {
        self.textureManager = textureManager
        let game = BaseSettings.game
        switch game.materialType {
        case .standard: _material = DefaultMaterial(textureManager: textureManager)
        }
    }
   
    public func buildMaterialFromProperties(_ mp: MaterialProps) -> SCNMaterial {
        return _material.buildMaterialFromProperties(mp)
    }

    func buildMaterial() -> SCNMaterial {
        return _material.buildMaterial()
    }

    func buildMaterialBlended(src sourceBlendMode: BlendMode, dst destinationBlendMode: BlendMode) -> SCNMaterial {
        return _material.buildMaterialBlended(src: sourceBlendMode, dst: destinationBlendMode)
    }

    func buildMaterialTested(cutoff: Float = 0.5) -> SCNMaterial {
        return _material.buildMaterialTested(cutoff: cutoff)
    }
}
